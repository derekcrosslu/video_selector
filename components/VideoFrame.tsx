import React, { useRef, useState, useEffect } from "react";
import styles from "./VideoFrame.module.css";
import VideoPlayer from "react-video-js-player";
// this is the only component that I tested besides the app component
// because jest runner needs additional logic to do the "end to end" test and without it
// some components will break for not being render below their parent component's and thus props appear as 


function VideoFrame(props: any) {
// custom video player
function onPlayerReady(player:any) {
  console.log("Player is ready: ", player);
}

function onVideoPlay(duration:any) {
  console.log("Video played at: ", duration);
}

function onVideoPause(duration:any) {
  console.log("Video paused at: ", duration);
}

function onVideoTimeUpdate(duration:any) {
  console.log("Time updated: ", duration);
}

function onVideoSeeking(duration:any) {
  console.log("Video seeking: ", duration);
}

function onVideoSeeked(from:any, to:any) {
  console.log(`Video seeked from ${from} to ${to}`);
}

function onVideoEnd() {
  console.log("Video ended");
}
  
    return (
      <main className={styles.main}>
        <VideoPlayer
          className={styles.video}
          bigPlayButton={true}
          autoplay={true}
          controls={true}
          src={props.url}
          onReady={onPlayerReady}
          onPlay={onVideoPlay}
          onPause={onVideoPause}
          onTimeUpdate={onVideoTimeUpdate}
          onSeeking={onVideoSeeking}
          onSeeked={onVideoSeeked}
          onEnd={onVideoEnd}
          data-testid="video-frame-component"
        />
      </main>
    );
}



export default VideoFrame;


