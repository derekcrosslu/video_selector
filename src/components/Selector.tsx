import React from 'react'
import styles from "./Player.module.css";
//disregard this component, this same code appears in player component
function Selector(props:any) {
    return (
        <div>
          <button
            onClick={(e) => props.videoIdHandler(e.target)}
            className={styles.selectorContainer}
          >
            {props.screens.Selector.map((view: any, index: number) => {
              return (
                <div>
                  <img
                    src={view.thumbnail}
                    alt={view.description}
                    key={index}
                    id={view.id}
                    className={styles.thumbnail}
                  />
                  <div className={styles.duration}>
                    <h3>{view.info} min</h3>
                  </div>
                </div>
              );
            })}
          </button>
        </div>
    )
}

export default Selector
