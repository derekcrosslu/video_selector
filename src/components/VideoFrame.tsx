import React from "react";
import styles from "./VideoFrame.module.css";

function VideoFrame(props: any) {
    return (
      <main className={styles.main}>
        <video className={styles.video} controls>
          <source
            src={props.url}
            type="video/mp4"
            data-testid="video-frame-component"
          />
        </video>
      </main>
    );
}

export default VideoFrame;


