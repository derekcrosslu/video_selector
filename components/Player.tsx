import React from "react";
import  VideoFrame  from "./VideoFrame";
import styles from "./Player.module.css";


function Player(props: any) {
 // dirty functional component for selecting video...its actually faster than pulling it from state
  // I can think of a better strategy for sure but this was the quickest for now
  const VideoFrameSelector = (props: any) => {

    if (props.video.video_id === "sintel")
      return <VideoFrame url={props.video.video_file_url} />;
    if (props.video.video_id === "bbb")
      return <VideoFrame url={props.video.video_file_url} />;
    if (props.video.video_id === "walle")
      return <VideoFrame url={props.video.video_file_url} />;
    else return <VideoFrame url={props.video.video_file_url} />;
  };
  //little snippets to generate component key
function uuid() {
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function (c) {
    var r = (Math.random() * 16) | 0,
      v = c === "x" ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}
  // I created a component Selector to grouping and testing the thumbnail selector code but decided
  // to put it back into the player component for complications with app state
  return (
    <div className={styles.container}>
      <main className={styles.main}>
        <h1 className={styles.title}>{props.PlayerElements.video_title}</h1>

        <p className={styles.description}>
          <code className={styles.text}>
            {props.PlayerElements.video_description}
          </code>
        </p>
        <div>
          <VideoFrameSelector
            video={props.PlayerElements}
            data-testid="video-frame-selector"
          />
        </div>

        <div className={styles.selectorContainer}>
          {props.screens.Selector.map((view: any, index: number) => {
            const player_key = uuid();
            return (
              <div key={player_key}>
                // fix linking bug: moved button wrapper to avoid covering the minutes
                <button onClick={(e) => props.videoIdHandler(e.target)}>
                  <img
                    src={view.thumbnail}
                    alt={view.description}
                    id={view.id}
                    className={styles.thumbnail}
                  />
                </button>
                <div className={styles.duration}>
                  <h3>{view.info} min</h3>
                </div>
              </div>
            );
          })}
        </div>
      </main>
    </div>
  );
}
export default Player;
