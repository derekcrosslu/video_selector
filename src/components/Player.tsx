import React from "react";
import  VideoFrame  from "./VideoFrame";
import styles from "./Player.module.css";
import Selector from "./Selector"

function Player(props: any) {

  const VideoFrameSelector = (props: any) => {

    if (props.video.video_id === "sintel")
      return <VideoFrame url={props.video.video_file_url} />;
    if (props.video.video_id === "bbb")
      return <VideoFrame url={props.video.video_file_url} />;
    if (props.video.video_id === "walle")
      return <VideoFrame url={props.video.video_file_url} />;
    else return <VideoFrame url={props.video.video_file_url} />;
  };

function uuid() {
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function (c) {
    var r = (Math.random() * 16) | 0,
      v = c === "x" ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}

  return (
    <div className={styles.container}>
      <main className={styles.main}>
        <h1 className={styles.title}>{props.PlayerElements.video_title}</h1>

        <p className={styles.description}>
          <code className={styles.text}>
            {props.PlayerElements.video_description}
          </code>
        </p>
        <div>
          <VideoFrameSelector
            video={props.PlayerElements}
            data-testid="video-frame-selector"
          />
        </div>

        <div>
          <button
            onClick={(e) => props.videoIdHandler(e.target)}
            className={styles.selectorContainer}
          >
            {props.screens.Selector.map((view: any, index: number) => {
               const player_key = uuid();
              return (
                <div>
                  <img
                    src={view.thumbnail}
                    alt={view.description}
                    key={player_key}
                    id={view.id}
                    className={styles.thumbnail}
                  />
                  <div className={styles.duration}>
                    <h3>{view.info} min</h3>
                  </div>
                </div>
              );
            })}
          </button>
        </div>
      </main>
    </div>
  );
}
export default Player;
