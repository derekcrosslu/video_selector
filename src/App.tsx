import { useState, useEffect } from "react";
import { Screens } from "./screens.model";
import Player from "./../components/Player";
import axios from "axios";

function App() {
  const [screens, setScreens] = useState<Screens>({
    Selector: [],
    BrandWrapper: [],
  });
  const [VideoId, SetVideoId] = useState("bbb");
  // Separating informative or Selector data and resource files (mp4 and png) or Branding data
  // before saving it to state after receiving it from the api
  // look into the interface for model criteria
  const screenFilterHandler = (ApiResponse: any) => {
    const Selector = ApiResponse.screens.filter(
      (screen: any) => screen.type === "Selector"
    )[0].config.views;
    const BrandWrapper = ApiResponse.screens.filter(
      (screen: any) => screen.type === "BrandWrapper"
    );
    setScreens({
      Selector: Selector,
      BrandWrapper: BrandWrapper,
    });
  };
  useEffect(() => {
    axios
      .get("http://localhost:9000")
      .then((res) => {
        screenFilterHandler(res.data);
      })
      .catch((err) => console.error(err));
  }, []);
  const [PlayerElements, SetPlayerElements] = useState({});
  // saving brand id or "video id"
  const videoIdHandler = (ev: any) => {
    SetVideoId(ev.id);
  };
  //filtering data related to thumbnail selected
  const selectedVideo = screens.BrandWrapper.filter(
    (video: any) => video.id === VideoId
  )[0];
  const selectedView = screens.Selector.filter(
    (view: any) => view.id === VideoId
  )[0];
  // saving elements selected by clicking the thumbnail to render video player
  useEffect(() => {
    SetPlayerElements({
      header_image_url: selectedVideo && selectedVideo.config.views[0].url,
      video_title: selectedView && selectedView.title,
      video_description: selectedView && selectedView.description,
      video_duration: selectedView && selectedView.info,
      video_thumbnail: selectedView && selectedView.thumbnail,
      video_file_url:
        selectedVideo &&
        selectedVideo.config.views.filter(
          (view: any) => view.type === "smPlayerIframe"
        )[0].files[0].url,
      video_id: VideoId,
    });
  }, [selectedVideo]);

  return (
    <div className="container">
      <Player
        screens={screens}
        PlayerElements={PlayerElements}
        videoIdHandler={videoIdHandler}
        VideoId={VideoId}
        data-testid="video-player"
      />
    </div>
  );
}

export default App;
