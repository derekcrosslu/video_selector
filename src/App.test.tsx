import * as React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import App from "./App";
import Selector from "./components/Selector";
import Player from "./components/Player";
import VideoFrame from "./components/VideoFrame";
import { createRenderer } from "react-dom/test-utils";
import "@testing-library/jest-dom";
import TestRenderer from 'react-test-renderer';


describe("Comparing snapshots for App component", () => {
  test("matches snapshot", () => {
    const tree = TestRenderer.create(<App  />).toJSON();
    expect(tree).toMatchSnapshot()
  });
});

describe("VideoFrame component renders correctly", () => {
  test("Test video source type from VideoFrame component", () => {
    const { queryByTestId } = render(<VideoFrame />);
    expect(queryByTestId("video-frame-component")).toBeTruthy();
  });
})