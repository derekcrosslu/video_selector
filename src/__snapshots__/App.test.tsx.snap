// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comparing snapshots for App component matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="container"
  >
    <main
      className="main"
    >
      <h1
        className="title"
      />
      <p
        className="description"
      >
        <code
          className="text"
        />
      </p>
      <div>
        <main
          className="main"
        >
          <video
            className="video"
            controls={true}
          >
            <source
              data-testid="video-frame-component"
              type="video/mp4"
            />
          </video>
        </main>
      </div>
      <div>
        <button
          className="selectorContainer"
          onClick={[Function]}
        />
      </div>
    </main>
  </div>
</div>
`;
